Make temporary permission changes of destination directory optional

andrew-farries opened the pull request:
As of https://github.com/otiai10/copy/pull/9 the `Copy` operation makes temporary changes to the permissions on the destination directory to allow for the case when the source directory is not writeable.

This behaviour is undesirable in the case where the destination directory is not owned by the process running the `Copy` - in that case the `chmod` operation will fail causing the copy to abort. This can happen in Kubernetes when mounting a volume into a container; the mount point will usually be owned by `root`. Using such a directory as the destination for a `Copy` is currently not possible using this module if the copying process is not also running as `root`.

This PR adds a new option, `NoTemporaryPermChanges`, to skip the temporary modification of permissions on the target directory.

otiai10 commented:
Can you add some test cases to describe what you want to achieve in this proposal?

andrew-farries commented:
I've added a test for the new option.

What I want to achieve with this proposal is described in the `README.md` for the test case:

```
The destination folder is not owned by the process running the `Copy` operation, but is
writeable by that process.

In this case, the Copy operation requires the `NoTemporaryPermChanges` option to be set, 
otherwise a `chmod` operation is attempted on the destination dir, which will fail.
```

andrew-farries commented:
Is there anything else we should add to this? In summary we have:

* Added a new option that ensures that a `Copy` works when copying to a `dst` dir that is writeable by, but not owned by, the process running the `Copy`.
* Added a test for the new option.
* Made the test containers run as non-root in order to be able to write the test.

otiai10 commented:
@andrew-farries Thank you for reminding. I need time to think about the best interface of `Options` a bit.

otiai10 commented:
Anyhow, I've got your point and really appreciate your commits. Thank you

otiai10 commented:
@andrew-farries Quick question, do your test cases correctly fail without your change?

andrew-farries commented:
For docker yes, the new test will fail without the setting the option.

For vagrant and the tests that run directly on the GH actions runner, the test will pass regardless of whether the new option is set or not.

The reason is that the 'root owned directory' setup is only performed by the docker tests at the moment (in 834180edb7763f58f11f90ac06c2a262d5a173a8).

otiai10 commented:
@andrew-farries Can you please check #70 whenever you have time ;)
